### PR TITLE
Disposable interface and a bunch of general resource containers.

### PR DESCRIPTION
**Pull request by akarnokd**
Mostly internal addition plus the public `Disposable` interface.

Naming of packages and classes can be changed at any time if one wishes...

### CODE REVIEW

File: src/main/java/io/reactivex/internal/disposables/ArrayCompositeResource.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+/**
+ * A composite resource with a fixed number of slots.
+ *
+ * <p>Note that since the implementation leaks the methods of AtomicReferenceArray, one must be
+ * careful to only call setResource, replaceResource and dispose on it. All other methods may lead to undefined behavior
+ * and should be used by internal means only.
+ * 
+ * @param <T> the resource tpye
+ */
+public final class ArrayCompositeResource<T> extends AtomicReferenceArray<Object> implements Disposable {
```
Comment (artem-zinnatullin) [inline src/main/java/io/reactivex/internal/disposables/ArrayCompositeResource.java]: Just curious: why you use inheritance over composition for such classes? I saw other classes in RxJava that extend `AtomicInteger`, etc. Why? :)

Comment (JakeWharton) [inline src/main/java/io/reactivex/internal/disposables/ArrayCompositeResource.java]: Saves bytes. Collapse two object headers into one.

Comment (artem-zinnatullin) [inline src/main/java/io/reactivex/internal/disposables/ArrayCompositeResource.java]: Sure, but exposes API. int vs enums 2.

Comment (JakeWharton) [inline src/main/java/io/reactivex/internal/disposables/ArrayCompositeResource.java]: This is not public API

Comment (artem-zinnatullin) [inline src/main/java/io/reactivex/internal/disposables/ArrayCompositeResource.java]: Yes, but 

mug-meme-okay

Comment (akarnokd) [inline src/main/java/io/reactivex/internal/disposables/ArrayCompositeResource.java]: It saves bytes, allocation and gives access to intrinsified atomic operations such as getAndIncrement and others.

**Pull request merged**